### PR TITLE
Fix #4287: Logical operators on connection filtering with interfaces

### LINKED ISSUE #4287: Interface relationship connection where logical operators does not filter the results in Cypher
**Describe the bug**
When using the logical operators on the relationship connection where filter when the relationship is to an Interface type, the generated Cypher does not contain the filters specified inside the logical operators.

**Type definitions**
```graphql
type Actor {
    name: String
    actedIn: [Production!]! @relationship(type: "ACTED_IN", properties: "actedIn", direction: OUT)
}
interface actedIn @relationshipProperties {
    role: String
}
interface Production {
    title: String
}
type Movie implements Production{
    title: String
    runtime: Int
}
type Series implements Production{
    title: String
    episodes: Int
}
```

**To Reproduce**
1. Execute the following Query:
```graphql
query {
    actors {
        actedInConnection(where:  {
          OR: [
            {
              node: {
                title: "something"
              },
            }
            {
              node: {
                title: "whatever"
              }
            }
          ]  
        }) {
          edges {
            node {
              __typename
              title
            }
          }
        } 
    }
}
```

2. Notice how the generated Cypher does not filter the results in any way (there is no `WHERE`):
```sql
MATCH (this:Actor)
CALL {
    WITH this
    CALL {
        WITH this
        MATCH (this)-[this0:ACTED_IN]->(this1:Movie)
        WITH { node: { __resolveType: "Movie", __id: id(this1), title: this1.title } } AS edge
        RETURN edge
        UNION
        WITH this
        MATCH (this)-[this2:ACTED_IN]->(this3:Series)
        WITH { node: { __resolveType: "Series", __id: id(this3), title: this3.title } } AS edge
        RETURN edge
    }
    WITH collect(edge) AS edges
    WITH edges, size(edges) AS totalCount
    RETURN { edges: edges, totalCount: totalCount } AS var4
}
RETURN this { actedInConnection: var4 } AS this
```

**Expected behavior**
The results should have title either "something" or "whatever".
This is a regression from 4.0, when the translation looked like this:
```sql
MATCH (this:Actor)
CALL {
    WITH this
    CALL {
        WITH this
        MATCH (this)-[this0:ACTED_IN]->(this1:Movie)
        WHERE (this1.title = "something" OR this1.title = "whatever")
        WITH { node: { __resolveType: "Movie", __id: id(this1), title: this1.title } } AS edge
        RETURN edge
        UNION
        WITH this
        MATCH (this)-[this2:ACTED_IN]->(this3:Series)
        WHERE (this3.title = "something" OR this3.title = "whatever")
        WITH { node: { __resolveType: "Series", __id: id(this3), title: this3.title } } AS edge
        RETURN edge
    }
    WITH collect(edge) AS edges
    WITH edges, size(edges) AS totalCount
    RETURN { edges: edges, totalCount: totalCount } AS var4
}
RETURN this { actedInConnection: var4 } AS this
```

### PR DESCRIPTION
**Pull request by angrykoala**
# Description

Fix logical operators on interface connection filtering

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
Made some changes to the Neo4j GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at https://neo4j-graphql-toolbox-4306.surge.sh

Use for instance a local dbms hosted on [Neo4j Desktop](https://neo4j.com/download/) or a cloud-hosted [Aura Free](https://neo4j.com/cloud/aura-free/) dbms.

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3404 | 3404 | 30 | 35 | 1134 |
|  aggregations.TopLevelAggregateWithMultipleFields | 6802 | 6802 | 38 | 44 | 1134 |
|  aggregations.NestedAggregation | 16553 | 16553 | 54 | 66 | 2174 |
|  aggregations.AggregationWithWhere | 11979 | 11979 | 35 | 48 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116987 | 22116987 | 2208 | 2306 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116987 | 22116987 | 2141 | 2263 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9734 | 9734 | 61 | 53 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9937 | 9937 | 43 | 58 | 2174 |
|  aggregations.DeeplyNestedCount | 11062443 | 11062443 | 2940 | 3065 | 2008534 |
|  aggregations.InterfacesAggregations | 6242 | 6242 | 39 | 46 | 2080 |
|  aggregations.InterfacesAggregationsWithTwoFields | 11444 | 11444 | 62 | 89 | 2080 |
|  batch-create.BatchCreate | 4200 | 4200 | 134 | 187 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 103 | 114 | 11 |
|  connect.createAndConnect | 14424 | 14424 | 148 | 167 | 3003 |
|  connections.Connection | 14082 | 14082 | 60 | 63 | 2174 |
|  connections.NestedConnection | 41118 | 41118 | 100 | 93 | 4516 |
|  create.SimpleMutation | 7 | 7 | 48 | 49 | 1 |
|  cypher-directive.TopLevelMutationDirective | 1135 | 1135 | 54 | 22 | 1134 |
|  delete.SimpleDelete | 19401 | 19401 | 218 | 187 | 1040 |
|  delete.NestedDeleteInUpdate | 19198 | 19198 | 168 | 185 | 2040 |
|  2871.NestedRelationshipFilter | 23123 | 23123 | 86 | 60 | 4436 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 32 | 44 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22900 | 22900 | 81 | 83 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 5201 | 5201 | 32 | 41 | 1040 |
|  2925.NestedSingleRelationshipRequiredFilter | 9361 | 9361 | 41 | 53 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 16 | 20 | 1040 |
|  query.SimpleQueryWithRelationship | 16162 | 16162 | 32 | 46 | 2174 |
|  query.QueryWhere | 9705 | 9706 | 28 | 36 | 2164 |
|  query.SimpleQueryWithNestedWhere | 9883 | 9884 | 38 | 49 | 2164 |
|  query.Nested | 10096041 | 10096041 | 6108 | 6772 | 2008534 |
|  query.NestedWithFilter | 10074401 | 10074401 | 6004 | 6829 | 2004000 |
|  query.OrFilterOnRelationships | 40645 | 42798 | 139 | 177 | 1181 |
|  query.OrFilterOnRelationshipsAndNested | 35000 | 35681 | 160 | 198 | 1181 |
|  query.QueryWithNestedIn | 13224 | 13364 | 41 | 49 | 1158 |
|  query.NestedConnectionWhere | 9834 | 9834 | 44 | 51 | 2174 |
|  query.DeeplyNestedConnectionWhere | 9893 | 9893 | 60 | 82 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 19146 | 19152 | 109 | 140 | 1598 |
|  query.NestedWithRelationshipSingleFilters | 3881 | 3881 | 145 | 179 | 1134 |
|  query.Fulltext | 80 | 80 | 26 | 39 | 16 |
|  query.FulltextWithNestedQuery | 587 | 587 | 33 | 41 | 84 |
|  sorting-and-cypher.TopLevelSortWithCypher | 14092 | 14092 | 47 | 57 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypher | 14092 | 14092 | 51 | 68 | 2174 |
|  sorting-and-cypher.TopLevelSortWithCypherWithNested | 14190 | 14190 | 57 | 66 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypherWithNested | 14190 | 14190 | 69 | 78 | 2174 |
|  sorting-and-cypher.TopLevelSortWithExpensiveCypher | 14629 | 14629 | 84 | 102 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithExpensiveCypher | 14473 | 14473 | 78 | 104 | 2174 |
|  sorting.SortMultipleTypes | 3515 | 3515 | 54 | 58 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 14580 | 14580 | 78 | 75 | 2174 |
|  sorting.SortOnNestedFields | 14082 | 14082 | 41 | 43 | 2174 |
|  sorting.SortDeeplyNestedFields | 43198 | 43198 | 65 | 87 | 4516 |
|  sorting.ConnectionWithSort | 3284 | 3284 | 46 | 55 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 46 | 53 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 61 | 58 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 306 | 231 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 215 | 270 | 33033 |
|  update.NestedUpdate | 16143 | 16143 | 82 | 106 | 2002 |

</details>
        
Old Schema Generation: 23.151s
Schema Generation: 23.521s
Old Subgraph Schema Generation: 27.209s
Subgraph Schema Generation: 26.681s

File: packages/graphql/src/translate/queryAST/factory/FilterFactory.ts [hunk 1]
```diff
@@ -510,4 +506,29 @@ export class FilterFactory {
             filters,
         });
     }
+
+    private getConcreteFilter(entity: EntityAdapter, where: Record<string, any>): Filter | undefined {
+        const concreteEntities = getConcreteEntities(entity);
+        const nodeFilters: Filter[] = [];
+        for (const concreteEntity of concreteEntities) {
+            const concreteEntityWhere: Record<string, any> = where[concreteEntity.name];
+            if (!concreteEntityWhere) continue;
```
Comment (mjfwebb) [inline packages/graphql/src/translate/queryAST/factory/FilterFactory.ts:515]:
```suggestion
            if (!concreteEntityWhere) {
                continue;
            }
```

File: packages/graphql/tests/integration/issues/4287.int.test.ts [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+import { Neo4jGraphQL } from "../../../src/classes";
+import { cleanNodes } from "../../utils/clean-nodes";
+import { UniqueType } from "../../utils/graphql-types";
+import Neo4j from "../neo4j";
+
+describe("https://github.com/neo4j/graphql/issues/4287", () => {
+    let driver: Driver;
+    let neo4j: Neo4j;
+    let neo4jGraphql: Neo4jGraphQL;
+    const Movie = new UniqueType("Movie");
```
Comment (mjfwebb) [inline packages/graphql/tests/integration/issues/4287.int.test.ts:32]: This doesn't seem to be really used, but it probably should be. Actor should also be using UniqueType

Comment (angrykoala) [reply packages/graphql/tests/integration/issues/4287.int.test.ts:32]: Good catch, forgot to change the typedefs to use uniquetype

**Review by mjfwebb — APPROVED** [review]

**Pull request merged**